File: features/step_definitions/subscription_steps.rb
Given(/^A user invited me to a ledger$/) do
  @inviter = User.where(email: "jane.doe@gmail.com").first_or_create
  builder = LedgerBuilder.make(@inviter)
  @invited_ledger = builder.create_ledger(currency: "USD")
  manager = LedgerManager.new(@invited_ledger, @inviter)
  manager.invite("john.doe@gmail.com")
  @subscription = Subscription.last
end

Given(/^the invite is used$/) do
  @subscription.update(done: true)
end

When(/^I invite "(.+)"$/) do |email_address|
  find_link("Menu").click
  find_link("Invite Subscribers").click
  within '#invite-section' do
    fill_in "Email", with: email_address
    click_button "Send Invitation"
  end
end

Then(/^the system should send an invite to "(.+)"$/) do |email_address|
  expect(page).to have_content("Invitation Sent")
  email = ActionMailer::Base.deliveries.first
  expect(email).not_to be_blank
  expect(email.to.first).to eql(email_address)
end

Then(/^I should see a link to subscribe to the ledger$/) do
  expect(page).to have_content("You have been invited to contribute to:")
  expect(page).to have_content(@invited_ledger.name)
  find_link("Subscribe")
end

When(/^I follow the ledger subscription link$/) do
  visit(subscription_path(@subscription))
end

Then(/^I should see the ledger$/) do
  expect(page).to have_content(@invited_ledger.name)
end

Then(/^I should be able to subscribe to the ledger$/) do
  find_link("Subscribe").click
  expect(page).to have_content(@invited_ledger.name)
  me = User.find_by(email:"john.doe@gmail.com")
  subscribed = me.subscribed_ledgers.find(@invited_ledger.id)
  expect(subscribed).not_to be_blank
end

File: spec/cypress/support/step_definitions/subscription_steps.js
import users from '../../fixtures/users';

Given(/^A user invited me to a ledger$/, () => {
  cy.pondoScript('invite', {
    sender: users.jessy,
    my_email: users.jane.email
  }).as('inviteData');
});

When(/^I invite '(.+)'$/, emailAddress => {
  cy.get('[href="#header-menu"]').click();
  cy.findLink('Invite Subscribers').click();
  cy.get('#invite-section').within(() => {
    cy.findField('Email').clear().type(emailAddress);
    cy.findButton('Send Invitation').click();
  });
});

When('I follow the ledger subscription link', () => {
  cy.get('@inviteData').then(data => {
    const { subscription } = data;
    cy.pondoRun(`subscription_path Subscription.find('${subscription.id}')`)
      .then(url => {
        cy.visit(url);
      });
  })
});

When('the invite is used', () => {
  cy.get('@inviteData').then(({ subscription }) => {
    cy.pondoRun(`Subscription.find('${subscription.id}').update(done: true)`);
  });
});

Then(/^the system should send an invite to '(.+)'$/, emailAddress => {
  cy.contains('Invitation Sent');
  cy.pondoScript('last_email_delivery').then(email => {
    cy.wrap(email).should('eql', emailAddress);
  });
});

Then('I should see a link to subscribe to the ledger', () => {
  cy.get('@inviteData').then(inviteData => {
    const { ledger } = inviteData;
    cy.get('body')
      .should('contain', 'You have been invited to contribute to:')
      .and('contain', ledger.name)
  });
  cy.findLink('Subscribe');
});

Then('I should be able to subscribe to the ledger', () => {
  cy.get('.body').within(() => {
    cy.findLink('Subscribe').click();
  });
  cy.get('@inviteData').then(inviteData => {
    const { ledger, subscription } = inviteData;
    cy.get('body').should('contain', ledger.name)
    cy.pondoScript(
      'find_user_subscription',
      {
        email: users.jane.email,
        ledger: ledger.id,
      }
    ).its('id').should('not.eql', null);
  });
});

Then('I should see the ledger', () => {
  cy.get('@inviteData').then(({ ledger }) => {
    cy.get('header').contains(ledger.name);
  });
});
